engine/namegen.ml

```diff
@@ -471,3 +471,34 @@ let rename_bound_vars_as_displayed sigma avoid env c =
     | _ -> c
   in
   rename avoid env c
+
+let u_ident = Id.of_string "u"
+
+let universe_binders_with_opt_names orig names =
+  let orig = Univ.AUContext.names orig in
+  let orig = Array.to_list orig in
+  let udecl = match names with
+  | None -> orig
+  | Some udecl ->
+    try
+      List.map2 (fun orig {CAst.v = na} ->
+          match na with
+          | Anonymous -> orig
+          | Name id -> Name id) orig udecl
+    with Invalid_argument _ ->
+      let len = List.length orig in
+      CErrors.user_err ~hdr:"universe_binders_with_opt_names"
+        Pp.(str "Universe instance should have length " ++ int len)
+  in
+  let fold_named i ubind = function
+    | Name id -> Id.Map.add id (Univ.Level.var i) ubind
+    | Anonymous -> ubind
+  in
+  let ubind = List.fold_left_i fold_named 0 UnivNames.empty_binders udecl in
+  let fold_anons i ubind = function
+    | Name _ -> ubind
+    | Anonymous ->
+      let id = next_ident_away_from u_ident (fun id -> Id.Map.mem id ubind) in
+      Id.Map.add id (Univ.Level.var i) ubind
+  in
+  List.fold_left_i fold_anons 0 ubind udecl
```

Comment:
This is quadratic in the number of anonymous universe variables.  Please do something like this instead (untested) to make it linear (or polylogarithmic, depending on the performance of `Id.Map.mem`))
```suggestion
  let fold_anons i (u_ident, ubind) = function
    | Name _ -> ubind
    | Anonymous ->
      let id = next_ident_away_from u_ident (fun id -> Id.Map.mem id ubind) in
      (id, Id.Map.add id (Univ.Level.var i) ubind)
  in
  let (_, ubind) = List.fold_left_i fold_anons 0 (u_ident, ubind) udecl in
  ubind
```